graphql/src/execution/ast.rs

```diff
@@ -0,0 +1,349 @@
+/// `SelectionSet` is constructed, fragments must already have been resolved
+/// as it only allows using fields.
+///
+/// The set of types that a `SelectionSet` can accommodate must be set at
+/// the time the `SelectionSet` is constructed. It is not possible to add
+/// more types to it, but it is possible to add fields for all known types
+/// or only some of them
+#[derive(Debug, Clone, PartialEq)]
+pub struct SelectionSet {
+    // Map object types to the list of fields that should be selected for
+    // them
+    items: Vec<(String, Vec<Field>)>,
+}
+
+impl SelectionSet {
+    /// Create a new `SelectionSet` that can handle the given types
+    pub fn new(types: Vec<String>) -> Self {
+        let items = types.into_iter().map(|name| (name, Vec::new())).collect();
+        SelectionSet { items }
+    }
+
+    /// Create a new `SelectionSet` that can handle the same types as
+    /// `other`, but ignore all fields from `other`
+    pub fn empty_from(other: &SelectionSet) -> Self {
+        let items = other
+            .items
+            .iter()
+            .map(|(name, _)| (name.clone(), Vec::new()))
+            .collect();
+        SelectionSet { items }
+    }
+
+    /// Return `true` if this selection set does not select any fields for
+    /// its types
+    pub fn is_empty(&self) -> bool {
+        self.items.iter().all(|(_, fields)| fields.is_empty())
+    }
+
+    /// If the selection set contains a single field across all its types,
+    /// return it. Otherwise, return `None`
+    pub fn single_field(&self) -> Option<&Field> {
+        let mut iter = self.items.iter();
+        let field = match iter.next() {
+            Some((_, fields)) => {
+                if fields.len() != 1 {
+                    return None;
+                } else {
+                    &fields[0]
+                }
+            }
+            None => return None,
+        };
+        for (_, fields) in iter {
+            if fields.len() != 1 {
+                return None;
+            }
+            if &fields[0] != field {
+                return None;
+            }
+        }
+        return Some(field);
+    }
+
+    /// Iterate over all types and the fields for those types
+    pub fn fields(&self) -> impl Iterator<Item = (&str, impl Iterator<Item = &Field>)> {
+        self.items
+            .iter()
+            .map(|(name, fields)| (name.as_str(), fields.iter()))
+    }
+
+    /// Iterate over all types and the fields that are not leaf fields, i.e.
+    /// whose selection sets are not empty
+    pub fn interior_fields(&self) -> impl Iterator<Item = (&str, impl Iterator<Item = &Field>)> {
+        self.items.iter().map(|(name, fields)| {
+            (
+                name.as_str(),
+                fields.iter().filter(|field| !field.is_leaf()),
+            )
+        })
+    }
+
+    /// Iterate over all fields for the given object type
+    pub fn fields_for(&self, obj_type: &s::ObjectType) -> impl Iterator<Item = &Field> {
```

Comment:
This should document the panic condition.

Reply:
Added